core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+use std::{collections::HashMap, fmt};
+
+use anyhow::Error;
+use async_trait::async_trait;
+use zksync_core::sync_layer::MainNodeClient;
+use zksync_dal::{ConnectionPool, SqlxError, StorageProcessor};
+use zksync_object_store::{ObjectStore, ObjectStoreError};
+use zksync_types::{
+    snapshots::{
+        SnapshotFactoryDependencies, SnapshotHeader, SnapshotRecoveryStatus, SnapshotStorageLog,
+        SnapshotStorageLogsChunk, SnapshotStorageLogsStorageKey,
+    },
+    zkevm_test_harness::k256::pkcs8::der::EncodeValue,
+    MiniblockNumber, H256,
+};
+use zksync_utils::bytecode::hash_bytecode;
+use zksync_web3_decl::{
+    jsonrpsee::core::ClientError as RpcError,
+    namespaces::{EthNamespaceClient, ZksNamespaceClient},
+};
+
+use crate::SnapshotApplierError::*;
+
+mod tests;
+
+#[derive(thiserror::Error, Debug)]
+pub enum SnapshotApplierError {
+    #[error("canceled")]
+    Canceled(String),
+    #[error(transparent)]
+    Fatal(#[from] anyhow::Error),
+    #[error("retryable")]
+    Retryable(String),
+}
+pub struct SnapshotsApplier<'a> {
+    connection_pool: &'a ConnectionPool,
+    blob_store: Box<dyn ObjectStore>,
+    applied_snapshot_status: SnapshotRecoveryStatus,
+}
+
+impl From<ObjectStoreError> for SnapshotApplierError {
+    fn from(error: ObjectStoreError) -> Self {
+        match error {
+            ObjectStoreError::KeyNotFound(err) => {
+                Fatal(Error::msg(format!("Key not found in object store: {err}")))
+            }
+            ObjectStoreError::Serialization(err) => {
+                Fatal(Error::msg(format!("Unable to deserialize object: {err}")))
+            }
+            ObjectStoreError::Other(_) => {
+                Retryable("An error occurred while accessing object store".to_string())
+            }
+        }
+    }
+}
+
+impl From<SqlxError> for SnapshotApplierError {
+    fn from(err: SqlxError) -> Self {
+        match err {
+            SqlxError::Database(_) => Fatal(err.into()),
+            SqlxError::RowNotFound => Fatal(err.into()),
+            SqlxError::ColumnNotFound(_) => Fatal(err.into()),
+            SqlxError::Configuration(_) => Fatal(err.into()),
+            SqlxError::TypeNotFound { type_name: _ } => Fatal(err.into()),
+            _ => Retryable(format!("An error occured when accessing DB: {err}")),
+        }
+    }
+}
+
+#[async_trait]
+pub trait SnapshotsApplierMainNodeClient: fmt::Debug + Send + Sync {
+    async fn miniblock_hash(&self, number: MiniblockNumber) -> Result<Option<H256>, RpcError>;
+
+    async fn fetch_newest_snapshot(&self) -> Result<Option<SnapshotHeader>, RpcError>;
+}
+impl<'a> SnapshotsApplier<'a> {
+    pub async fn load_snapshot(
```

Comment:
Would this be the entry point used by EN code? I'd expect `SnapshotApplierError` to be an internal detail (cf. updated EN components with similar transient / fatal error distinction, such as the consistency checker), so that retries are handled internally in the snapshot applier, and it only returns fatal errors / cancellations outside. It could be the case that the retry strategy should be configurable in this case, since otherwise we risk getting indefinitely stuck if we improperly classify a fatal error as transient.

Reply:
I plan to refactor the code to behave similarly as you described in resiliency PR